<a href="https://colab.research.google.com/github/proteus21/DATA-SCIENCE-STUDY/blob/main/Machine%20Learning/02_regression/01_linear_regression_normal_equation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### scikit-learn
Strona biblioteki: [https://scikit-learn.org](https://scikit-learn.org)  

Dokumentacja/User Guide: [https://scikit-learn.org/stable/user_guide.html](https://scikit-learn.org/stable/user_guide.html)

Podstawowa biblioteka do uczenia maszynowego w języku Python.

Aby zainstalować bibliotekę scikit-learn, użyj polecenia poniżej:
```
!pip install scikit-learn
```
Aby zaktualizować do najnowszej wersji bibliotekę scikit-learn, użyj polecenia poniżej:
```
!pip install --upgrade scikit-learn
```
Kurs stworzony w oparciu o wersję `0.22.1`

### Spis treści:
1. [Import bibliotek](#0)
2. [Przygotowanie danych](#1)
3. [Równanie normalne](#2)
4. [Regresja liniowa przy pomocy scikit-learn](#3)




### <a name='0'></a> Import bibliotek

In [ ]:
import numpy as np

### <a name='1'></a> Przygotowanie danych

In [ ]:
X1 = np.array([1, 2, 3, 4, 5, 6])
Y = np.array([3000, 3250, 3500, 3750, 4000, 4250])
m = len(X1)

print(f'Lata pracy: {X1}')
print(f'Wynagrodzenie: {Y}')
print(f'Liczba próbek: {m}')

Lata pracy: [1 2 3 4 5 6]
Wynagrodzenie: [3000 3250 3500 3750 4000 4250]
Liczba próbek: 6


In [ ]:
X1 = X1.reshape(m, 1) #zmiana rozmiaru 6 wiersz i jedna kolumna
print(X1)
print(X1.shape)

[[1]
 [2]
 [3]
 [4]
 [5]
 [6]]
(6, 1)


In [ ]:
bias = np.ones((m, 1)) # wygenerujemy wektor wyrazow wolnych
print(bias)
print(bias.shape)

[[1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]]
(6, 1)


In [ ]:
X = np.append(bias, X1, axis=1) # wektor wyrazow wolnych dodamy do kolumny X1
print(X)
print(X.shape)

[[1. 1.]
 [1. 2.]
 [1. 3.]
 [1. 4.]
 [1. 5.]
 [1. 6.]]
(6, 2)


### <a name='2'></a> Równanie normalne

Regresja liniowa w $R^2$:
$$Y = w_0 + w_1X_1$$
$$Y = W^TX$$
gdzie: $$W= \left[\begin{matrix}w_0\\w_1\end{matrix}\right]$$ $$ X= \left[\begin{matrix}1\\X_1\end{matrix}\right] $$ stąd $$ W^T= \left[\begin{matrix}w_0&w_1\end{matrix}\right] $$    
$$Y = W^TX = \left[\begin{matrix}w_0&w_1\end{matrix}\right] \cdot  \left[\begin{matrix}1\\X_1\end{matrix}\right] = w_0 + w_1X_1$$

Równanie normalne - równanie pozwalające obliczyć minimum funkcji straty (o ile istnieje)
$$W = (X^TX)^{-1}(X^TY)$$

In [ ]:
np.dot(X.T, X) # Transponowanie macierzy

array([[ 6., 21.],
       [21., 91.]])

In [ ]:
L = np.linalg.inv(np.dot(X.T, X)) #macierz odwrotna
L

array([[ 0.86666667, -0.2       ],
       [-0.2       ,  0.05714286]])

In [ ]:
P = np.dot(X.T, Y) #Transponowana macierz razy Y
P

array([21750., 80500.])

In [ ]:
np.dot(L, P) #

array([2750.,  250.])

Końcowa postać modelu$$Y = 2750 + 250X_1$$

### <a name='3'></a> Regresja liniowa przy pomocy scikit-learn

In [ ]:
from sklearn.linear_model import LinearRegression

regression = LinearRegression()
regression.fit(X1, Y)

print(regression.intercept_)
print(regression.coef_[0])

2750.0
250.0
